include/sys/onoff.h

```diff
@@ -0,0 +1,603 @@
+	onoff_service_transition_fn start;
+
+	/* Function to invoke to transition the service to off. */
+	onoff_service_transition_fn stop;
+
+	/* Function to force the service state to reset, where
+	 * supported.
+	 */
+	onoff_service_transition_fn reset;
+
+	/* Mutex protection for flags, clients, releaser, and refs. */
+	struct k_spinlock lock;
+
+	/* Client to be informed when transition to off completes. */
+	struct onoff_client *releaser;
+
+	/* Flags identifying the service state. */
+	u16_t flags;
+
+	/* Number of active clients for the service. */
+	u16_t refs;
+};
+
+/** @internal */
+#define ONOFF_SERVICE_INITIALIZER(_start, _stop, _reset, _flags) { \
+		.start = _start,				   \
+		.stop = _stop,					   \
+		.reset = _reset,				   \
+		.flags = _flags,				   \
+}
+
+/**
+ * @brief Initialize an on-off service to off state.
+ *
+ * This function must be invoked exactly once per service instance, by
+ * the infrastructure that provides the service, and before any other
+ * on-off service API is invoked on the service.
+ *
+ * This function should never be invoked by clients of an on-off service.
+ *
+ * @param srv the service definition object to be initialized.
+ *
+ * @param start the function used to (initiate a) transition from off
+ * to on.  This must not be null.  Include @ref ONOFF_SERVICE_START_SLEEPS as
+ * appropriate in flags.
+ *
+ * @param stop the function used to (initiate a) transition from on to
+ * off.  This must not be null.  Include @ref ONOFF_SERVICE_STOP_SLEEPS
+ * as appropriate in flags.
+ *
+ * @param reset the function used to clear errors and force the
+ * service to an off state.  Pass null if the service cannot or need
+ * not be reset.  (Services where a transition operation can complete
+ * with an error notification should support the reset operation.)
+ * Include @ref ONOFF_SERVICE_RESET_SLEEPS as appropriate in flags.
+ *
+ * @param flags any or all of the flags mentioned above,
+ * e.g. @ref ONOFF_SERVICE_START_SLEEPS.  Use of other flags produces an
+ * error.
+ *
+ * @retval 0 on success
+ * @retval -EINVAL if start, stop, or flags are invalid
+ */
+int onoff_service_init(struct onoff_service *srv,
+		       onoff_service_transition_fn start,
+		       onoff_service_transition_fn stop,
+		       onoff_service_transition_fn reset,
+		       u32_t flags);
+
+/** @internal
+ *
+ * Flag fields used to specify on-off client behavior.
+ *
+ * These flags control whether calls to onoff_service_request() and
+ * onoff_service_release() are synchronous or asynchronous, and for
+ * asynchronous operations how the operation result is communicated to
+ * the client.
+ */
+enum onoff_client_flags {
+	/* Known-invalid field, used in validation */
+	ONOFF_CLIENT_NOTIFY_INVALID     = 0,
+
+	/*
+	 * Indicates that no notification will be provided.
+	 *
+	 * Callers must check for completions using
+	 * onoff_client_fetch_result().
+	 *
+	 * See async_operation_init_spinwait().
```

Comment:
onoff_client_init_spinwait(). You could add @ref.

Reply:
Fixed; thanks.

Adding `@ref` is not necessary and has no effect: writing it as a function call `foo()` would cause Doxygen to attempt to generate a link to the corresponding documented element.  That doesn't show up in Zephyr's generated documentation because it would require dropping into `@rst` mode and using a `:c:func:` prefix, which only works on whole paragraphs.